File: examples/csharp/src/examples/Example05Dx11MovingTriangle.cs
using DXBridgeExamples.DxBridge;

namespace DXBridgeExamples.Examples;

internal static class Example05Dx11MovingTriangle
{
    private const string VsSource =
        "struct VSInput { float2 pos : POSITION; float3 col : COLOR; };\n"
        + "struct PSInput { float4 pos : SV_POSITION; float3 col : COLOR; };\n"
        + "PSInput main(VSInput v) {\n"
        + "    PSInput o;\n"
        + "    o.pos = float4(v.pos, 0.0f, 1.0f);\n"
        + "    o.col = v.col;\n"
        + "    return o;\n"
        + "}\n";

    private const string PsSource =
        "struct PSInput { float4 pos : SV_POSITION; float3 col : COLOR; };\n"
        + "float4 main(PSInput p) : SV_TARGET {\n"
        + "    return float4(p.col, 1.0f);\n"
        + "}\n";

    public static int Run(string[] args)
    {
        var cli = new CommandLine(
            args,
            new[] { "--dll", "--width", "--height", "--frames", "--speed", "--amplitude", "--sync-interval" },
            new[] { "--hidden", "--debug" });
        if (cli.HelpRequested)
        {
            Console.WriteLine("Usage: Example05Dx11MovingTriangle [--dll PATH] [--width 960] [--height 540] [--frames 0] [--speed 1.6] [--amplitude 0.38] [--hidden] [--debug] [--sync-interval 1]");
            return 0;
        }

        var width = cli.GetInt("--width", 960);
        var height = cli.GetInt("--height", 540);
        var frames = cli.GetInt("--frames", 0);
        var speed = cli.GetFloat("--speed", 1.6f);
        var amplitude = cli.GetFloat("--amplitude", 0.38f);
        var syncInterval = cli.GetInt("--sync-interval", 1);

        using var window = new Win32Window("DXBridge C# Moving Triangle", width, height, !cli.HasFlag("--hidden"), "DXBridgeCSharpMovingTriangle");
        using var bridge = new DXBridge(cli.GetString("--dll"));

        var device = DXBridge.NullHandle;
        var swapChain = DXBridge.NullHandle;
        var rtv = DXBridge.NullHandle;
        var vertexShader = DXBridge.NullHandle;
        var pixelShader = DXBridge.NullHandle;
        var inputLayout = DXBridge.NullHandle;
        var pipeline = DXBridge.NullHandle;
        var vertexBuffer = DXBridge.NullHandle;
        var initialized = false;

        try
        {
            window.Create();
            Console.WriteLine($"Loaded: {bridge.DllPath}");
            Console.WriteLine($"Created HWND: 0x{window.Handle.ToInt64():X16}");

            bridge.Init(DXBridge.DxbBackendDx11);
            initialized = true;

            device = bridge.CreateDevice(DXBridge.DxbBackendDx11, 0, cli.HasFlag("--debug") ? DXBridge.DxbDeviceFlagDebug : 0u);
            swapChain = bridge.CreateSwapChain(device, window.Handle, width, height);
            var backBuffer = bridge.GetBackBuffer(swapChain);
            rtv = bridge.CreateRenderTargetView(device, backBuffer);

            vertexShader = bridge.CompileShader(device, DXBridge.DxbShaderStageVertex, VsSource, "moving_triangle_vs", "main", "vs_5_0", DXBridge.DxbCompileDebug);
            pixelShader = bridge.CompileShader(device, DXBridge.DxbShaderStagePixel, PsSource, "moving_triangle_ps", "main", "ps_5_0", DXBridge.DxbCompileDebug);

            var elements = new[]
            {
                new DXBInputElementDesc
                {
                    StructVersion = DXBridge.StructVersion,
                    Semantic = DXBridge.DxbSemanticPosition,
                    SemanticIndex = 0,
                    Format = DXBridge.DxbFormatRg32Float,
                    InputSlot = 0,
                    ByteOffset = 0,
                },
                new DXBInputElementDesc
                {
                    StructVersion = DXBridge.StructVersion,
                    Semantic = DXBridge.DxbSemanticColor,
                    SemanticIndex = 0,
                    Format = DXBridge.DxbFormatRgb32Float,
                    InputSlot = 0,
                    ByteOffset = 8,
                },
            };

            inputLayout = bridge.CreateInputLayout(device, elements, vertexShader);
            pipeline = bridge.CreatePipelineState(device, vertexShader, pixelShader, inputLayout);
            vertexBuffer = bridge.CreateBuffer(device, DXBridge.DxbBufferFlagVertex | DXBridge.DxbBufferFlagDynamic, 15 * sizeof(float), 5 * sizeof(float));

            var startTime = DateTime.UtcNow;
            var frameIndex = 0;
            Console.WriteLine("Rendering triangle animation. Close the window to exit.");

            while (window.PumpMessages())
            {
                var elapsed = (DateTime.UtcNow - startTime).TotalSeconds;
                var offsetX = (float)(amplitude * Math.Sin(elapsed * speed));
                var offsetY = (float)(0.06 * Math.Sin(elapsed * speed * 1.9));

                bridge.UploadFloats(vertexBuffer, MakeVertexData(offsetX, offsetY));

                bridge.BeginFrame(device);
                bridge.SetRenderTargets(device, rtv);
                bridge.ClearRenderTarget(rtv, new[] { 0.04f, 0.05f, 0.08f, 1.0f });
                bridge.SetViewport(device, 0.0f, 0.0f, width, height, 0.0f, 1.0f);
                bridge.SetPipeline(device, pipeline);
                bridge.SetVertexBuffer(device, vertexBuffer, 5 * sizeof(float), 0);
                bridge.Draw(device, 3, 0);
                bridge.EndFrame(device);
                bridge.Present(swapChain, syncInterval);

                frameIndex++;
                if (frames > 0 && frameIndex >= frames)
                {
                    window.Destroy();
                    break;
                }
            }

            Console.WriteLine($"Done. Rendered frames: {frameIndex}");
        }
        finally
        {
            if (vertexBuffer != DXBridge.NullHandle)
            {
                bridge.DestroyBuffer(vertexBuffer);
            }

            if (pipeline != DXBridge.NullHandle)
            {
                bridge.DestroyPipeline(pipeline);
            }

            if (inputLayout != DXBridge.NullHandle)
            {
                bridge.DestroyInputLayout(inputLayout);
            }

            if (pixelShader != DXBridge.NullHandle)
            {
                bridge.DestroyShader(pixelShader);
            }

            if (vertexShader != DXBridge.NullHandle)
            {
                bridge.DestroyShader(vertexShader);
            }

            if (rtv != DXBridge.NullHandle)
            {
                bridge.DestroyRenderTargetView(rtv);
            }

            if (swapChain != DXBridge.NullHandle)
            {
                bridge.DestroySwapChain(swapChain);
            }

            if (device != DXBridge.NullHandle)
            {
                bridge.DestroyDevice(device);
            }

            if (initialized)
            {
                bridge.Shutdown();
            }
        }

        return 0;
    }

    private static float[] MakeVertexData(float offsetX, float offsetY)
    {
        return
        [
            0.0f + offsetX, 0.52f + offsetY, 1.0f, 0.1f, 0.1f,
            -0.42f + offsetX, -0.38f + offsetY, 0.1f, 1.0f, 0.1f,
            0.42f + offsetX, -0.38f + offsetY, 0.1f, 0.3f, 1.0f,
        ];
    }
}

File: examples/java/src/examples/Example05Dx11MovingTriangle.java
package examples;

import dxbridge.CommandLine;
import dxbridge.DXBridge;
import dxbridge.DXBridgeLibrary;
import dxbridge.Win32Window;

public final class Example05Dx11MovingTriangle {
    private static final String VS_SOURCE =
        "struct VSInput { float2 pos : POSITION; float3 col : COLOR; };\n"
        + "struct PSInput { float4 pos : SV_POSITION; float3 col : COLOR; };\n"
        + "PSInput main(VSInput v) {\n"
        + "    PSInput o;\n"
        + "    o.pos = float4(v.pos, 0.0f, 1.0f);\n"
        + "    o.col = v.col;\n"
        + "    return o;\n"
        + "}\n";

    private static final String PS_SOURCE =
        "struct PSInput { float4 pos : SV_POSITION; float3 col : COLOR; };\n"
        + "float4 main(PSInput p) : SV_TARGET {\n"
        + "    return float4(p.col, 1.0f);\n"
        + "}\n";

    public static void main(String[] args) {
        CommandLine cli = parse(args);
        if (cli.isHelpRequested()) {
            printUsage();
            return;
        }

        int width = cli.getInt("--width", 960);
        int height = cli.getInt("--height", 540);
        int frames = cli.getInt("--frames", 0);
        float speed = cli.getFloat("--speed", 1.6f);
        float amplitude = cli.getFloat("--amplitude", 0.38f);
        int syncInterval = cli.getInt("--sync-interval", 1);

        Win32Window window = new Win32Window(
            "DXBridge Java Moving Triangle",
            width,
            height,
            !cli.hasFlag("--hidden"),
            "DXBridgeJavaMovingTriangle"
        );
        DXBridge bridge = new DXBridge(cli.getString("--dll", null));

        long device = DXBridge.DXBRIDGE_NULL_HANDLE;
        long swapChain = DXBridge.DXBRIDGE_NULL_HANDLE;
        long rtv = DXBridge.DXBRIDGE_NULL_HANDLE;
        long vertexShader = DXBridge.DXBRIDGE_NULL_HANDLE;
        long pixelShader = DXBridge.DXBRIDGE_NULL_HANDLE;
        long inputLayout = DXBridge.DXBRIDGE_NULL_HANDLE;
        long pipeline = DXBridge.DXBRIDGE_NULL_HANDLE;
        long vertexBuffer = DXBridge.DXBRIDGE_NULL_HANDLE;
        boolean initialized = false;

        try {
            window.create();
            System.out.println("Loaded: " + bridge.getDllPath().getAbsolutePath());
            System.out.printf("Created HWND: 0x%016X%n", window.getHwndValue());

            bridge.init(DXBridge.DXB_BACKEND_DX11);
            initialized = true;

            device = bridge.createDevice(
                DXBridge.DXB_BACKEND_DX11,
                0,
                cli.hasFlag("--debug") ? DXBridge.DXB_DEVICE_FLAG_DEBUG : 0
            );
            swapChain = bridge.createSwapChain(device, window.getHwndPointer(), width, height, DXBridge.DXB_FORMAT_RGBA8_UNORM, 2, 0);
            long backBuffer = bridge.getBackBuffer(swapChain);
            rtv = bridge.createRenderTargetView(device, backBuffer);

            vertexShader = bridge.compileShader(
                device,
                DXBridge.DXB_SHADER_STAGE_VERTEX,
                VS_SOURCE,
                "moving_triangle_vs",
                "main",
                "vs_5_0",
                DXBridge.DXB_COMPILE_DEBUG
            );
            pixelShader = bridge.compileShader(
                device,
                DXBridge.DXB_SHADER_STAGE_PIXEL,
                PS_SOURCE,
                "moving_triangle_ps",
                "main",
                "ps_5_0",
                DXBridge.DXB_COMPILE_DEBUG
            );

            DXBridgeLibrary.DXBInputElementDesc[] elements = DXBridge.newInputElementArray(2);
            elements[0].semantic = DXBridge.DXB_SEMANTIC_POSITION;
            elements[0].semantic_index = 0;
            elements[0].format = DXBridge.DXB_FORMAT_RG32_FLOAT;
            elements[0].input_slot = 0;
            elements[0].byte_offset = 0;

            elements[1].semantic = DXBridge.DXB_SEMANTIC_COLOR;
            elements[1].semantic_index = 0;
            elements[1].format = DXBridge.DXB_FORMAT_RGB32_FLOAT;
            elements[1].input_slot = 0;
            elements[1].byte_offset = 8;

            inputLayout = bridge.createInputLayout(device, elements, vertexShader);
            pipeline = bridge.createPipelineState(
                device,
                vertexShader,
                pixelShader,
                inputLayout,
                DXBridge.DXB_PRIM_TRIANGLELIST,
                0,
                0,
                0,
                0
            );
            vertexBuffer = bridge.createBuffer(
                device,
                DXBridge.DXB_BUFFER_FLAG_VERTEX | DXBridge.DXB_BUFFER_FLAG_DYNAMIC,
                15 * 4,
                5 * 4
            );

            long startTime = System.nanoTime();
            int frameIndex = 0;
            System.out.println("Rendering triangle animation. Close the window to exit.");

            while (window.pumpMessages()) {
                double elapsed = (System.nanoTime() - startTime) / 1000000000.0;
                float offsetX = (float) (amplitude * Math.sin(elapsed * speed));
                float offsetY = (float) (0.06 * Math.sin(elapsed * speed * 1.9));

                bridge.uploadFloats(vertexBuffer, makeVertexData(offsetX, offsetY));

                bridge.beginFrame(device);
                bridge.setRenderTargets(device, rtv, DXBridge.DXBRIDGE_NULL_HANDLE);
                bridge.clearRenderTarget(rtv, new float[] { 0.04f, 0.05f, 0.08f, 1.0f });
                bridge.setViewport(device, 0.0f, 0.0f, (float) width, (float) height, 0.0f, 1.0f);
                bridge.setPipeline(device, pipeline);
                bridge.setVertexBuffer(device, vertexBuffer, 5 * 4, 0);
                bridge.draw(device, 3, 0);
                bridge.endFrame(device);
                bridge.present(swapChain, Math.max(0, syncInterval));

                frameIndex++;
                if (frames > 0 && frameIndex >= frames) {
                    window.destroy();
                    break;
                }
            }

            System.out.println("Done. Rendered frames: " + frameIndex);
        } finally {
            if (vertexBuffer != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyBuffer(vertexBuffer);
            }
            if (pipeline != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyPipeline(pipeline);
            }
            if (inputLayout != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyInputLayout(inputLayout);
            }
            if (pixelShader != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyShader(pixelShader);
            }
            if (vertexShader != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyShader(vertexShader);
            }
            if (rtv != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyRenderTargetView(rtv);
            }
            if (swapChain != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroySwapChain(swapChain);
            }
            if (device != DXBridge.DXBRIDGE_NULL_HANDLE) {
                bridge.destroyDevice(device);
            }
            if (initialized) {
                bridge.shutdown();
            }
            window.destroy();
        }
    }

    private static float[] makeVertexData(float offsetX, float offsetY) {
        return new float[] {
            0.0f + offsetX, 0.52f + offsetY, 1.0f, 0.1f, 0.1f,
            -0.42f + offsetX, -0.38f + offsetY, 0.1f, 1.0f, 0.1f,
            0.42f + offsetX, -0.38f + offsetY, 0.1f, 0.3f, 1.0f
        };
    }

    private static CommandLine parse(String[] args) {
        return new CommandLine(
            args,
            CommandLine.options("--dll", "--width", "--height", "--frames", "--speed", "--amplitude", "--sync-interval"),
            CommandLine.options("--hidden", "--debug")
        );
    }

    private static void printUsage() {
        System.out.println(
            "Usage: Example05Dx11MovingTriangle [--dll PATH] [--width 960] [--height 540] [--frames 0] [--speed 1.6] [--amplitude 0.38] [--hidden] [--debug] [--sync-interval 1]"
        );
    }
}
